File: benchmarks/erlang/string_processing.erl
-module(string_processing).
-export([main/0]).

process_strings(Count, Size) ->
    process_strings(Count, Size, "").

process_strings(0, _, Result) ->
    Result;
process_strings(Count, Size, Result) ->
    Str = create_random_string(Size),
    Processed = process_string(Str),
    process_strings(Count - 1, Size, Result ++ Processed).

create_random_string(Size) ->
    Chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789",
    CharsLength = length(Chars),
    create_random_string(Size, Chars, CharsLength, []).

create_random_string(0, _, _, Result) ->
    lists:reverse(Result);
create_random_string(Size, Chars, CharsLength, Result) ->
    RandomIndex = rand:uniform(CharsLength) - 1,
    RandomChar = lists:nth(RandomIndex + 1, Chars),
    create_random_string(Size - 1, Chars, CharsLength, [RandomChar | Result]).

process_string(Input) ->
    % Replace all vowels with their uppercase version
    Result1 = lists:foldl(
        fun({From, To}, Acc) ->
            re:replace(Acc, From, To, [global, {return, list}])
        end,
        Input,
        [{"a", "A"}, {"e", "E"}, {"i", "I"}, {"o", "O"}, {"u", "U"}]
    ),
    
    % Replace all digits with their doubled value
    Result2 = lists:foldl(
        fun(I, Acc) ->
            Digit = integer_to_list(I),
            Doubled = integer_to_list(I * 2),
            re:replace(Acc, Digit, Doubled, [global, {return, list}])
        end,
        Result1,
        lists:seq(0, 9)
    ),
    
    % Capitalize the first letter
    Result3 = case Result2 of
        [] -> [];
        [First | Rest] ->
            [string:to_upper(First) | Rest]
    end,
    
    % Reverse the string
    Reversed = lists:reverse(Result3),
    
    % Take the first half of the reversed string
    HalfLen = length(Reversed) div 2,
    string:slice(Reversed, 0, HalfLen).

main() ->
    % Seed the random number generator
    rand:seed(exsss, {erlang:monotonic_time(), erlang:time_offset(), erlang:unique_integer()}),
    StartTime = erlang:system_time(millisecond),
    
    % Process strings of different sizes
    Small = process_strings(10000, 10),  % 10,000 strings of length 10
    Medium = process_strings(1000, 100), % 1,000 strings of length 100
    Large = process_strings(100, 1000),  % 100 strings of length 1,000
    
    ResultLength = length(Small) + length(Medium) + length(Large),
    io:format("Processed string length: ~p~n", [ResultLength]),
    
    ElapsedTime = erlang:system_time(millisecond) - StartTime,
    io:format("Time taken: ~p ms~n", [ElapsedTime]),
    
    {ok, [TotalWords, _]} = process_info(self(), memory),
    io:format("Memory used: ~p KB~n", [TotalWords div 1024]),
    ok.

File: benchmarks/clojure/string_processing.clj
(ns string-processing-benchmark
  (:require [clojure.string :as str])
  (:gen-class))

(defn create-random-string [size]
  (let [chars "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"]
    (apply str (repeatedly size #(rand-nth chars)))))

(defn process-string [input]
  (let [; Replace all vowels with their uppercase version
        result (-> input
                 (str/replace "a" "A")
                 (str/replace "e" "E")
                 (str/replace "i" "I")
                 (str/replace "o" "O")
                 (str/replace "u" "U"))
        
        ; Replace all digits with their doubled value
        result (reduce (fn [s i]
                         (str/replace s (str i) (str (* i 2))))
                       result
                       (range 10))
        
        ; Capitalize the first letter
        result (if (> (count result) 0)
                 (str/upper-case (subs result 0 1)) 
                 "") 
                 
        ; Rest of the string (excluding first letter)
        result (if (> (count input) 1)
                 (str result (subs input 1))
                 result)
                 
        ; Reverse the string
        reversed (apply str (reverse result))
        
        ; Take the first half of the reversed string
        half-len (quot (count reversed) 2)]
    
    (subs reversed 0 half-len)))

(defn process-strings [count size]
  (let [result (atom "")]
    (dotimes [i count]
      (let [str (create-random-string size)
            processed (process-string str)]
        (swap! result str processed)))
    @result))

(defn -main [& args]
  (let [start-time (System/currentTimeMillis)]
    
    ; Process strings of different sizes
    (let [small (process-strings 10000 10)    ; 10,000 strings of length 10
          medium (process-strings 1000 100)   ; 1,000 strings of length 100
          large (process-strings 100 1000)    ; 100 strings of length 1,000
          
          result-length (+ (count small) (count medium) (count large))]
      
      (println (str "Processed string length: " result-length))
      
      (println (str "Time taken: " (- (System/currentTimeMillis) start-time) " ms"))
      
      ; Estimate memory usage
      (println (str "Memory used: " (/ (- (Runtime/getRuntime) (.freeMemory (Runtime/getRuntime))) 1024) " KB")))))